proto/fields.py

```diff
@@ -155,6 +261,153 @@ def __init__(self, key_type, value_type, *, number: int, message=None, enum=None
+            # no such values at all
+            if field.oneof != self.field.oneof:
+                continue
+
+            delattr(instance, field_name)
+
+    def __set__(self, instance, value):
+        """Called whenever a value is assigned to a proto.Field attribute on an instantiated
+        proto.Message object.
+
+        Usage:
+
+            class MyMessage(proto.Message):
+                name = proto.Field(proto.STRING, number=1)
+
+            my_message = MyMessage()
+            my_message.name = "Frodo"
+
+        In the above scenario, `__set__` is called with "Frodo" passed as `value` and `my_instance`
+        passed as `instance`.
+        """
+        value = self._set_coercion(value)
+        value = self._hydrate_dicts(value)
+
+        # Warning: `always_commit` is hacky!
+        # Some contexts, particularly instances created from MapFields, require immediate syncing.
+        # It is impossible to deduce such a scenario purely from logic available to this function,
+        # so instead we set a flag on instances when a MapField yields them, and then when those
+        # instances receive attribute updates, immediately syncing those values to the underlying
+        # pb2 instance is sufficient.
+        always_commit: bool = getattr(instance, '_always_commit', False)
+        if always_commit or not self.field.can_set_natively(value):
+            pb_value = instance._meta.marshal.to_proto(self.field.pb_type, value)
+            _pb = instance._meta.pb(**{self.original_name: pb_value})
+            instance._pb.ClearField(self.original_name)
+            instance._pb.MergeFrom(_pb)
+        else:
+
+            if value is None:
+                self.__delete__(instance)
+                instance._pb.MergeFrom(instance._meta._pb(**{self.original_name: None}))
+                return
+
+            instance._meta.marshal.validate_primitives(self.field, value)
+            instance._mark_pb_stale(self.original_name)
+
+        setattr(instance, self.instance_attr_name, value)
+        self._clear_oneofs(instance)
+
+    def __get__(self, instance: 'proto.Message', _):  # type: ignore
+        """Called whenever a value is read from a proto.Field attribute on an instantiated
+        proto.Message object.
+
+        Usage:
+
+            class MyMessage(proto.Message):
+                name = proto.Field(proto.STRING, number=1)
+
+            my_message = MyMessage(name="Frodo")
+            print(my_message.name)
+
+        In the above scenario, `__get__` is called with "my_message" passed as `instance`.
+        """
+        # If `instance` is None, then we are accessing this field directly
+        # off the class itself instead of off an instance.
+        if instance is None:
+            return self.original_name
+
+        value = getattr(instance, self.instance_attr_name, None)
+        if self.field.can_get_natively and value is not None:
+            return value
```

Comment:
I don't have enough context here to know whether `instance` can ever have an actual attr for the field with value `None`:  my first inclination was to suggest a marker such as `_nonesuch`:
```suggestion
        value = getattr(instance, self.instance_attr_name, _nonesuch)
        if self.field.can_get_natively and value is not _nonesuch:
            return value
```